### PR TITLE
Add LABEL annotation and label node to BOX annotation 

### PR DESCRIPTION
**Pull request by stockiNail**
Refactoring to reduce the duplicate code of PR #473 and #477

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,18580 +1,8 @@
 {
   "name": "chartjs-plugin-annotation",
   "version": "1.0.2",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Comment (kurkle) [inline package-lock.json:4]: very old npm?

File: docs/samples/types/box.md [hunk 2]
```diff
@@ -1,118 +1,123 @@
...
-    }
-  }
-];
-
-module.exports = {
-  actions: actions,
-  config: config,
-};
-```
+# Box
```
Comment (kurkle) [inline docs/samples/types/box.md:1]: Line ending change (could be valid, did not check)

File: test/fixtures/box/label.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
+module.exports = {
```
Comment (kurkle) [inline test/fixtures/box/label.js:1]: Looking at the fixture, I think the label should be clipped inside the border (not draw halfway over it)

File: test/fixtures/label/borderDash.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            text2: {
+              type: 'label',
+              xScaleID: 'x',
+              yScaleID: 'y',
+              xValue: 'February',
+              yValue: 11,
+              backgroundColor: 'rgba(101, 33, 171, 0.5)',
+              borderColor: 'rgb(101, 33, 171)',
+              borderDash: [6, 6],
+              borderWidth: 1,
```
Comment (kurkle) [inline test/fixtures/label/borderDash.js:40]: Should make the border wider for this test, so any regression is more likely catched

File: test/fixtures/label/borderRadius.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            },
+            text2: {
+              type: 'label',
+              xScaleID: 'x',
+              yScaleID: 'y',
+              xValue: 'February',
+              yValue: 11,
+              backgroundColor: 'rgb(245,245,245)',
+              borderDash: [6, 6],
+              borderWidth: 1,
```
Comment (kurkle) [inline test/fixtures/label/borderRadius.js:38]: Same here, wider border (and not dash) would make this test more likely to work

File: test/fixtures/label/contentMultiline.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+          annotations: {
+            text1: {
+              type: 'label',
+              xScaleID: 'x',
+              yScaleID: 'y',
+              xValue: 'January',
+              yValue: 20,
+              backgroundColor: 'transparent',
+              borderWidth: 0,
+              content: ['This is my text, row 1', 'This is my text, row 2', 'This is my text, row 3'],
```
Comment (kurkle) [inline test/fixtures/label/contentMultiline.js:27]: Should have different length lines for this test

File: test/fixtures/label/position.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          annotations: {
+            text1: {
+              type: 'label',
+              xScaleID: 'x',
+              yScaleID: 'y',
+              xValue: 'January',
+              yValue: 20,
+              backgroundColor: 'transparent',
+              borderWidth: 0,
+              content: 'This is my text'
```
Comment (kurkle) [inline test/fixtures/label/position.js:27]: Could use the position value as label, for example: `'poistion: undefined'`
Could also place all on same value and maybe make this a line chart with a line on that value, so there would be a reference for the eye.

I'm not 100% `position` is the correct property name for this functionality?

File: types/options.d.ts [hunk 8]
```diff
@@ -1,75 +1,81 @@
...
-  xValue?: Scriptable<ScaleValue, PartialEventContext>;
-  yValue?: Scriptable<ScaleValue, PartialEventContext>;
-}
-
-export interface AnnotationPluginOptions extends AnnotationEvents {
-  annotations: AnnotationOptions[] | Record<string, AnnotationOptions>,
-  dblClickSpeed?: Scriptable<number, PartialEventContext>,
-  drawTime?: Scriptable<DrawTime, PartialEventContext>,
-  animations: Record<string, unknown>,
+import { Color } from 'chart.js';
```
Comment (kurkle) [inline types/options.d.ts:1]: looks like another line end change, is it valid? I guess the linter should validate these.

File: src/types/labelItem.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    return widthCache.get(mapKey);
+  }
+}
+
+function getImageSize(size, value) {
+  if (typeof value === 'number') {
+    return value;
+  } else if (typeof value === 'string') {
+    return toPercent(value) * size;
```
Comment (kurkle) [inline src/types/labelItem.js:70]: A string not ending with '%' would end up being undefined?

**Review by kurkle — COMMENTED** [review]
Couple of random comments, will review the actual pattern later.

File: package-lock.json (hunk 1 above)
Comment (stockiNail) [reply package-lock.json:4]: my fault. I'm gonna replace with the master version

File: src/types/labelItem.js (hunk 9 above)
Comment (stockiNail) [reply src/types/labelItem.js:70]: No, it isn't. When a string is not ending with % (or is not a string) the function returns false (therefore 0).
Anyway this function is used in the line annotation for long time.

**Comment by stockiNail** [comment]
The line annotation is not touched at the moment. Nevertheless it should changed in order to leverage on Label object.
One of codeClimate issues is related to duplication of `getImage` function which could be solved when also LineAnnotation will use Label.

**Comment by kurkle** [comment]
| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `dist/chartjs-plugin-annotation.esm.js` | 8.77 kB | +1.17 kB (+15%) | ⚠️ |
| `dist/chartjs-plugin-annotation.js` | 8.9 kB | +1.18 kB (+15%) | ⚠️ |
| `dist/chartjs-plugin-annotation.min.js` | 6.3 kB | +873 B (+16%) | ⚠️ |

**Pull request closed without merging**